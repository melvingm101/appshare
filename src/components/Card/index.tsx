import React from "react";
import Link from "next/link";
import { CurrentProject } from "@/client/models";
import Image from "next/image";
import Pill from "../Pill";
import EmojiPicker from "../EmojiPicker";
import { useStore } from "@/zustand";
import ReactionList from "../ReactionList";

const Card = ({ project }: { project: CurrentProject }) => {
  const user = useStore((state) => state.user);
  return (
    <div className="flex flex-col bg-primary-color rounded-lg shadow mt-3 h-[310px]">
      <div className="px-4 flex flex-col justify-between h-full">
        <Link href={`/projects/${project.id}`}>
          <div>
            <div className="flex flex-col justify-between pt-4 pb-2 leading-normal hover:underline">
              <h5
                className="text-lg font-bold text-gray-900 dark:text-white text-left whitespace-nowrap overflow-hidden text-ellipsis"
                data-testid="card-title"
              >
                {project.title}
              </h5>
              <div className="text-xs">{`${project.views} ${
                project.views === 1 ? "view" : "views"
              } | ${project.likes.length} ${
                project.likes.length === 1 ? "reaction" : "reactions"
              }`}</div>
            </div>
            {project.banner ? (
              <div
                className="w-full h-[130px] rounded-sm mb-3"
                data-testid="card-banner"
              >
                <Image
                  alt={project.title}
                  src={project.banner}
                  width="0"
                  height="0"
                  sizes="100vw"
                  className="w-full h-full object-cover rounded-md"
                />
              </div>
            ) : (
              <div className="h-[130px] mb-3 overflow-hidden text-ellipsis text-sm">
                {project.description}
              </div>
            )}
          </div>
          <div className="flex flex-wrap">
            {project.tags.map((tag) => (
              <Pill tag={tag} key={tag} />
            ))}
          </div>
        </Link>
        <div className="flex">
          {user && <EmojiPicker id={project.id} isSinglePage={false} />}
          <ReactionList likes={project.likes} isSinglePage={false} />
        </div>
      </div>
    </div>
  );
};

export default Card;
