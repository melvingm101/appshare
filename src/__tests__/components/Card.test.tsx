import { render } from "@testing-library/react";
import Card from "../../components/Card";

describe("Card", () => {
  const project = {
    title: "Hello world",
    description: "Hello world",
    id: 1,
    views: 0,
    likes: [],
    projectUrl: "",
    tags: ["DESIGN"],
    banner: "",
  };

  it("renders a card component", () => {
    const { container } = render(<Card project={project} />);

    expect(container).toMatchSnapshot();
  });

  it("renders banner if banner is provided", () => {
    const newProject = { ...project, banner: "https://www.google.com" };
    const { getByTestId } = render(<Card project={newProject} />);

    expect(getByTestId("card-banner")).toBeTruthy();
  });
});
