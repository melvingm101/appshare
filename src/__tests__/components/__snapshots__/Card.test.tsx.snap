// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card renders a card component 1`] = `
<div>
  <div
    class="flex flex-col bg-primary-color rounded-lg shadow mt-3"
  >
    <div
      class="px-4"
    >
      <a
        href="/projects/1"
      >
        <div
          class="hover:underline"
        >
          <div
            class="flex flex-col justify-between pt-4 pb-2 leading-normal"
          >
            <h5
              class="mb-2 text-lg font-bold text-gray-900 dark:text-white text-left whitespace-nowrap overflow-hidden text-ellipsis"
              data-testid="card-title"
            >
              Hello world
            </h5>
          </div>
          <div
            class="h-[160px] mb-3 overflow-hidden text-ellipsis text-sm"
          >
            Hello world
          </div>
        </div>
        <div>
          <span
            class="text-xs font-medium mr-2 px-2.5 py-0.5 rounded-full bg-green-900 text-green-300"
          >
            Design
          </span>
        </div>
      </a>
      <div
        class="flex text-sm text-gray-500 mr-2"
      >
        <div
          class="mt-2 py-0.5 px-2 bg-secondary-color rounded-md flex items-center"
        >
          <svg
            aria-hidden="true"
            class="h-4 w-4 mr-1"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M2.036 12.322a1.012 1.012 0 010-.639C3.423 7.51 7.36 4.5 12 4.5c4.638 0 8.573 3.007 9.963 7.178.07.207.07.431 0 .639C20.577 16.49 16.64 19.5 12 19.5c-4.638 0-8.573-3.007-9.963-7.178z"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
            <path
              d="M15 12a3 3 0 11-6 0 3 3 0 016 0z"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
          0
        </div>
      </div>
      <div
        class="flex mb-3"
      >
        <div
          class="flex text-sm text-gray-500 mr-2"
        >
          <div
            class="mt-2 py-0.5 px-2 bg-secondary-color rounded-md flex items-center hover:text-white cursor-pointer"
          >
            <svg
              aria-hidden="true"
              class="h-4 w-4 mr-1"
              fill="none"
              stroke="currentColor"
              stroke-width="1.5"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M6.633 10.5c.806 0 1.533-.446 2.031-1.08a9.041 9.041 0 012.861-2.4c.723-.384 1.35-.956 1.653-1.715a4.498 4.498 0 00.322-1.672V3a.75.75 0 01.75-.75A2.25 2.25 0 0116.5 4.5c0 1.152-.26 2.243-.723 3.218-.266.558.107 1.282.725 1.282h3.126c1.026 0 1.945.694 2.054 1.715.045.422.068.85.068 1.285a11.95 11.95 0 01-2.649 7.521c-.388.482-.987.729-1.605.729H13.48c-.483 0-.964-.078-1.423-.23l-3.114-1.04a4.501 4.501 0 00-1.423-.23H5.904M14.25 9h2.25M5.904 18.75c.083.205.173.405.27.602.197.4-.078.898-.523.898h-.908c-.889 0-1.713-.518-1.972-1.368a12 12 0 01-.521-3.507c0-1.553.295-3.036.831-4.398C3.387 10.203 4.167 9.75 5 9.75h1.053c.472 0 .745.556.5.96a8.958 8.958 0 00-1.302 4.665c0 1.194.232 2.333.654 3.375z"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
            0
          </div>
        </div>
        <div
          class="flex text-sm text-gray-500 mr-2"
        >
          <div
            class="mt-2 py-0.5 px-2 bg-secondary-color rounded-md flex items-center hover:text-white cursor-pointer"
          >
            <svg
              aria-hidden="true"
              class="h-4 w-4 mr-1"
              fill="none"
              stroke="currentColor"
              stroke-width="1.5"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M7.5 15h2.25m8.024-9.75c.011.05.028.1.052.148.591 1.2.924 2.55.924 3.977a8.96 8.96 0 01-.999 4.125m.023-8.25c-.076-.365.183-.75.575-.75h.908c.889 0 1.713.518 1.972 1.368.339 1.11.521 2.287.521 3.507 0 1.553-.295 3.036-.831 4.398C20.613 14.547 19.833 15 19 15h-1.053c-.472 0-.745-.556-.5-.96a8.95 8.95 0 00.303-.54m.023-8.25H16.48a4.5 4.5 0 01-1.423-.23l-3.114-1.04a4.5 4.5 0 00-1.423-.23H6.504c-.618 0-1.217.247-1.605.729A11.95 11.95 0 002.25 12c0 .434.023.863.068 1.285C2.427 14.306 3.346 15 4.372 15h3.126c.618 0 .991.724.725 1.282A7.471 7.471 0 007.5 19.5a2.25 2.25 0 002.25 2.25.75.75 0 00.75-.75v-.633c0-.573.11-1.14.322-1.672.304-.76.93-1.33 1.653-1.715a9.04 9.04 0 002.86-2.4c.498-.634 1.226-1.08 2.032-1.08h.384"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
            0
          </div>
        </div>
        <div
          class="flex text-sm text-gray-500 mr-2"
        >
          <div
            class="mt-2 py-0.5 px-2 bg-secondary-color rounded-md flex items-center hover:text-white cursor-pointer"
          >
            <svg
              aria-hidden="true"
              class="h-4 w-4 mr-1"
              fill="none"
              stroke="currentColor"
              stroke-width="1.5"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M21 8.25c0-2.485-2.099-4.5-4.688-4.5-1.935 0-3.597 1.126-4.312 2.733-.715-1.607-2.377-2.733-4.313-2.733C5.1 3.75 3 5.765 3 8.25c0 7.22 9 12 9 12s9-4.78 9-12z"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
            0
          </div>
        </div>
        <div
          class="flex text-sm text-gray-500 mr-2"
        >
          <div
            class="mt-2 py-0.5 px-2 bg-secondary-color rounded-md flex items-center hover:text-white cursor-pointer"
          >
            <svg
              aria-hidden="true"
              class="h-4 w-4 mr-1"
              fill="none"
              stroke="currentColor"
              stroke-width="1.5"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M15.182 15.182a4.5 4.5 0 01-6.364 0M21 12a9 9 0 11-18 0 9 9 0 0118 0zM9.75 9.75c0 .414-.168.75-.375.75S9 10.164 9 9.75 9.168 9 9.375 9s.375.336.375.75zm-.375 0h.008v.015h-.008V9.75zm5.625 0c0 .414-.168.75-.375.75s-.375-.336-.375-.75.168-.75.375-.75.375.336.375.75zm-.375 0h.008v.015h-.008V9.75z"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
            0
          </div>
        </div>
        <div
          class="flex text-sm text-gray-500 mr-2"
        >
          <div
            class="mt-2 py-0.5 px-2 bg-secondary-color rounded-md flex items-center hover:text-white cursor-pointer"
          >
            <svg
              aria-hidden="true"
              class="h-4 w-4 mr-1"
              fill="none"
              stroke="currentColor"
              stroke-width="1.5"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M15.362 5.214A8.252 8.252 0 0112 21 8.25 8.25 0 016.038 7.048 8.287 8.287 0 009 9.6a8.983 8.983 0 013.361-6.867 8.21 8.21 0 003 2.48z"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
              <path
                d="M12 18a3.75 3.75 0 00.495-7.467 5.99 5.99 0 00-1.925 3.546 5.974 5.974 0 01-2.133-1A3.75 3.75 0 0012 18z"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
            0
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
